cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/NamespaceSourceControlMetadataRefresher.java

```diff
@@ -0,0 +1,210 @@
+  private final AtomicLong lastRefreshTime;
+
+  /**
+   * Constructs a {@link NamespaceSourceControlMetadataRefresher} instance.
+   */
+  public NamespaceSourceControlMetadataRefresher(TransactionRunner transactionRunner,
+      SourceControlOperationRunner sourceControlOperationRunner,
+      NamespaceId namespaceId) {
+    this.transactionRunner = transactionRunner;
+    this.sourceControlOperationRunner = sourceControlOperationRunner;
+    this.namespaceId = namespaceId;
+    this.lastRefreshTime = new AtomicLong(0);
+  }
+
+  private RepositorySourceControlMetadataStore getRepoSourceControlMetadataStore(
+      StructuredTableContext context) {
+    return RepositorySourceControlMetadataStore.create(context);
+  }
+
+  private NamespaceSourceControlMetadataStore getNamespaceSourceControlMetadataStore(
+      StructuredTableContext context) {
+    return NamespaceSourceControlMetadataStore.create(context);
+  }
+
+  private RepositoryTable getRepositoryTable(StructuredTableContext context)
+      throws TableNotFoundException {
+    return new RepositoryTable(context);
+  }
+
+
+  // TODO(CDAP-21020): Refactor this function
+  private RepositoryMeta getRepositoryMeta(NamespaceId namespace)
+      throws RepositoryNotFoundException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      RepositoryTable table = getRepositoryTable(context);
+      RepositoryMeta repoMeta = table.get(namespace);
+      if (repoMeta == null) {
+        throw new RepositoryNotFoundException(namespace);
+      }
+
+      return repoMeta;
+    }, RepositoryNotFoundException.class);
+  }
+
+  public long refreshMetadata() throws NotFoundException {
+    LOG.info(
+        "Started NamespaceSourceControlMetadataRefreshService manual refresh for namespace {}",
+        namespaceId);
+    long refreshStartTime = System.currentTimeMillis();
+
+    // Getting repository config for a specific namespace
+    RepositoryConfig repoConfig = getRepositoryMeta(namespaceId).getConfig();
+    // Listing the apps from remote repo
+    RepositoryAppsResponse repositoryAppsResponse = sourceControlOperationRunner.list(
+        new NamespaceRepository(namespaceId, repoConfig));
+
+    Set<String> repoFileNames = repositoryAppsResponse.getApps().stream()
+        .map(RepositoryApp::getName)
+        .collect(Collectors.toCollection(HashSet::new));
+
+    LOG.info(
+        "Done listing apps from repo in NamespaceSourceControlMetadataRefreshService manual refresh for namespace {} with repo size {}",
+        namespaceId, repoFileNames.size());
+
+    TransactionRunners.run(transactionRunner, context -> {
+      RepositorySourceControlMetadataStore repoMetadataStore = getRepoSourceControlMetadataStore(
+          context);
+
+      // Cleaning up the repo source control metadata table
+      cleanupRepoSourceControlMetaInBatchesWithNoTransaction(namespaceId.getNamespace(),
+          repoFileNames, repoMetadataStore);
+
+      LOG.info(
```

Comment:
change all info to debug

Reply:
I have removed the log because I thought logs like "Starting.." or "Finishing" might increase the log size. Any specific logs that should be added?

Reply:
no. looks good